### PR TITLE
simulation endpoint that retains state

### PR DESCRIPTION
**Pull request by ape-dev-cs**
This implements two new endpoints in the transaction simulator: 

- `simulate-stateful-new` which instantiates a new instance of an EVM simulation endpoint that can be used multiple times, with each new request simulating an vector of `SimulationRequest``. Returns a simulation id.
- `simulate-stateful/{id}` which takes the id provided in the URL and applies the array of `SimulationRequest` to the state of the EVM referred to by the id.

Uses a concurrent hashmap for in-memory EVM storage for max performance.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -36,12 +49,43 @@ pub fn simulate_bundle(
         .and_then(simulation::simulate_bundle)
 }
 
+/// POST /simulate-stateful-new
```
Comment (devanoneth) [inline src/lib.rs:52]:
```suggestion
/// POST /simulate-stateful
```

Would prefer just a POST without an id to setup your state. Would be more RESTful then.

File: src/lib.rs [hunk 2]
```diff
@@ -36,12 +49,43 @@ pub fn simulate_bundle(
...
+pub fn simulate_stateful_new(
+    config: Config,
+    state: Arc<SharedSimulationState>,
+) -> impl Filter<Extract = (impl Reply,), Error = Rejection> + Clone {
+    warp::path!("simulate-stateful-new")
```
Comment (devanoneth) [inline src/lib.rs:57]:
```suggestion
    warp::path!("simulate-stateful")
```

File: src/simulation.rs [hunk 3]
```diff
@@ -53,6 +58,22 @@ pub struct SimulationResponse {
...
+    #[serde(rename = "gasLimit")]
+    pub gas_limit: u64,
+    #[serde(rename = "blockNumber")]
+    pub block_number: Option<u64>,
+}
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
+pub struct StatefulSimulationResponse {
+    #[serde(rename = "statefulSimulationId")]
+    pub stateful_simulation_id: u32,
```
Comment (devanoneth) [inline src/simulation.rs:74]: I think UUIDs could be nicer than an incrementing id int. It means that if in future we allow a DB to be hooked up and multiple simulators to be run, the API doesn't need to change. I also frankly think the URLs will look nicer. :D  What do you think?

**Review by devanoneth — CHANGES_REQUESTED** [review]
Thanks for this ser, really nice! 👑 On the CI, if you pull in the latest changes to `main`, it should work now. RPC was timing out.

Could you also update the README with details on these new endpoints? Thanks!

**Comment by devanoneth** [comment]
Thanks for the contribution!

**Comment by ConnorSB13** [comment]
Thanks Mr Ape!

**Comment by polyphia** [comment]
Amazing !

**Pull request merged**